### PR TITLE
Update qml.generator's "hamiltonian" method to follow new operator arithmetic

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
We are generally treating "Hamiltonian" as meaning alternately `qml.ops.Hamiltonian` or `qml.ops.LinearCombination` depending on whether or not `new_opmath` is enabled. 

The current implementation of `qml.generator(op, format="hamiltonian")` mostly respects that, but uses `convert_to_legacy_H` if the generator is a `Sum`, `Prod` or `SProd` even with `new_opmath` is enabled, breaking the convention.

**Description of the Change:**
We add `convert_to_H`, which converts any arithmetic operator to a conceptual Hamiltonian - either `qml.ops.Hamiltonian` or `qml.ops.LinearCombination`. We use the new `convert_to_H` function in the `qml.generator` method. The `convert_to_legacy_H` function continues to convert specifically to a legacy Hamiltonian regardless of the global setting for `new_opmath`.

**Benefits:**
Consistency, no one will ask for a generator with `new_opmath` enabled and get a legacy Hamiltonian.

### CODE REVIEW

File: doc/development/deprecations.rst [hunk 1]
```diff
@@ -34,6 +34,7 @@ Pending deprecations
 
   - Added and deprecated for ``Sum`` and ``Prod`` instances in v0.35
 
+
```
Comment (lillian542) [inline doc/development/deprecations.rst:37]:
```suggestion
```

File: pennylane/operation.py [hunk 2]
```diff
@@ -3064,10 +3064,11 @@ def enable_new_opmath_cm():
 
 
 # pylint: disable=too-many-branches
-def convert_to_legacy_H(op):
+def convert_to_H(op):
```
Comment (lillian542) [inline pennylane/operation.py:3067]: Don't panic - `convert_to_legacy_H` is redefined below as a way to call this function with new_opmath temporarily disabled. Here the logic is generalized to make a version that will convert to a LinearCombination or Hamiltonian.

File: tests/test_operation.py [hunk 3]
```diff
@@ -2924,28 +2925,7 @@ def test_use_legacy_opmath_fixture():
             Tensor(qml.Z(1), qml.Z(2)),
         ],
     ),
-]
-
-
-@pytest.mark.parametrize("coeffs, obs", CONVERT_HAMILTONAIN)
-def test_convert_to_hamiltonian(coeffs, obs):
-    """Test that arithmetic operators can be converted to Hamiltonian instances"""
```
Comment (lillian542) [inline tests/test_operation.py:2932]: The diff here looks weird, but all that happened was the failing case in the xfail test below was moved up into the CONVERT_HAMILTONIAN list. This test is now in line 2945, and hasn't been modified.

**Comment by lillian542** [comment]
[sc-57982]

File: pennylane/operation.py [hunk 4]
```diff
@@ -3129,8 +3132,25 @@ def convert_to_legacy_H(op):
     else:
         raise ValueError("Could not convert to Hamiltonian. Some or all observables are not valid.")
 
+    return qml.Hamiltonian(coeffs, ops)
+
+
+def convert_to_legacy_H(op):
+    """
+    Converts arithmetic operators into :class:`~pennylane.Hamiltonian` instance.
```
Comment (mudit2812) [inline pennylane/operation.py:3140]:
```suggestion
    Converts arithmetic operators into a legacy :class:`~pennylane.Hamiltonian` instance.
```

**Review by mudit2812 — APPROVED** [review]

**Review by Qottmann — COMMENTED** [review]
I am in favor of this change, i.e. enforcing that generators are `Hamiltonian`-like instances for consistency

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -149,6 +149,7 @@
   additional functionality not available using legacy operators.
   [(#5253)](https://github.com/PennyLaneAI/pennylane/pull/5253)
   [(#5410)](https://github.com/PennyLaneAI/pennylane/pull/5410)
+  [(#5421)](https://github.com/PennyLaneAI/pennylane/pull/5410)
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:152]:
```suggestion
  [(#5421)](https://github.com/PennyLaneAI/pennylane/pull/5421)
```

Comment (lillian542) [reply doc/releases/changelog-dev.md:152]: 🙈

**Comment by lillian542** [comment]
> I am in favor of this change, i.e. enforcing that generators are Hamiltonian-like instances for consistency

I wouldn't say we're enforcing anything per se. This changes it specifically if you call `qml.generator(op, format="hamiltonian")`. Calls of `qml.generator` in general will still default to the `prefactor` format.

File: pennylane/operation.py (hunk 2 above)
Comment (PietropaoloFrisoni) [reply pennylane/operation.py:3067]: Very happy with this change

**Comment by PietropaoloFrisoni** [comment]
Thank you so much for this @lillian542 ! Tests seem to have issues, but it is most certainly just random fluctuations

**Comment by PietropaoloFrisoni** [comment]
@Qottmann Yep, I have the same issue in my PR (which is why I am still waiting to merge).

**Pull request merged**